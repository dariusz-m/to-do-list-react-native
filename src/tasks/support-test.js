// import ReactTestUtils from 'react-dom/test-utils';
// import ReactDOM from 'react-dom';
//
// /**
//  * @param {function} callback - Callback
//  * @param {string|number} firstParameter - First parameter
//  * @param {string|number} secondParameter - Second parameter
//  */
// export const expectedCallbackWasCalledOnceWith = (
//     callback, firstParameter = undefined, secondParameter = undefined,
// ) => {
//     expect(callback.mock.calls.length).toBe(1);
//     expect(callback.mock.calls[0][0]).toEqual(firstParameter);
//     expect(callback.mock.calls[0][1]).toEqual(secondParameter);
// };
//
// /**
//  * @param {React.Component} component - Component
//  * @param {string} buttonId - Button id
//  */
// export const clickButton = (component, buttonId) => {
//     const node = ReactDOM.findDOMNode(component);
//     const button = node.querySelector(`#${buttonId}`);
//     ReactTestUtils.Simulate.click(button);
// };
//
// /**
//  * Change input value
//  * @param {React.Component} component - Component
//  * @param {string} inputId - Input id
//  * @param {string|number} value - Value
//  */
// export const changeInputValue = (component, inputId, value) => {
//     const node = ReactDOM.findDOMNode(component);
//     const field = node.querySelector(`#${inputId}`);
//
//     field.value = value;
//     ReactTestUtils.Simulate.change(field);
// };
