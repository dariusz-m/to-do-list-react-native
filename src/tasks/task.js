import React from 'react';
import PropTypes from 'prop-types';

/**
 * Task component
 */
export class Task extends React.PureComponent {
    static propTypes = {
        task: PropTypes.shape({
            id: PropTypes.number.isRequired,
            task: PropTypes.string.isRequired,
        }).isRequired,
        deleteTask: PropTypes.func.isRequired,
        editTask: PropTypes.func.isRequired,
        updateTask: PropTypes.func.isRequired,
        cancelEditingTask: PropTypes.func.isRequired,
        changeEditingTask: PropTypes.func.isRequired,
        isEditing: PropTypes.bool.isRequired,
    };

    /**
     * Render
     * @returns {string}
     */
    render() {
        return (
            <li className="list-group-item">
                <button
                    className="btn btn-danger float-right"
                    onClick={() => this.props.deleteTask(this.props.task.id)}
                    id="delete-task-button"
                >
                    Delete
                </button>
                {
                    this.props.isEditing
                        ? <React.Fragment>
                            <input
                                className="pr-3"
                                value={this.props.task.task}
                                onChange={event => this.props.changeEditingTask(event.target.value)}
                            />
                            <button
                                className="btn btn-primary float-right mr-2"
                                onClick={() => this.props.cancelEditingTask()}>
                                Cancel
                            </button>
                            <button
                                className="btn btn-warning float-right mr-2"
                                id="update-task-button"
                                onClick={() => this.props.updateTask(this.props.task.id, this.props.task.task)}>
                                Update
                            </button>
                        </React.Fragment>
                        : <React.Fragment>
                            <p className="pr-3">{this.props.task.task}</p>
                            <button
                                className="btn btn-info float-right mr-2"
                                onClick={() => this.props.editTask(this.props.task.id)}>
                                Edit
                            </button>
                        </React.Fragment>
                }
            </li>
        );
    }
}
