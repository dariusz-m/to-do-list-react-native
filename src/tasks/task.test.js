import React from 'react';
import renderer from 'react-test-renderer';
import ReactTestUtils from 'react-dom/test-utils';
import { Task } from './task';
import { clickButton, expectedCallbackWasCalledOnceWith } from './support-test';

describe('Task tests', () => {
    it('Task component renders correctly without editing mode', () => {
        const task = { id: 6, task: 'Go running!' };
        const isEditing = false;
        const tree = renderer.create(
            <Task
                task={task}
                editTask={jest.fn()}
                cancelEditingTask={jest.fn()}
                deleteTask={jest.fn()}
                changeEditingTask={jest.fn()}
                updateTask={jest.fn()}
                isEditing={isEditing}
            />,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('Task component renders correctly with editing mode', () => {
        const task = { id: 6, task: 'Go running!' };
        const isEditing = true;
        const tree = renderer.create(
            <Task
                task={task}
                editTask={jest.fn()}
                cancelEditingTask={jest.fn()}
                deleteTask={jest.fn()}
                changeEditingTask={jest.fn()}
                updateTask={jest.fn()}
                isEditing={isEditing}
            />,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('After clicking the button to delete a task, callback(deleteTask) is called out once', () => {
        const task = { id: 3, task: 'Go running!' };
        const deleteTask = jest.fn();
        const component = ReactTestUtils.renderIntoDocument(
            <Task
                task={task}
                editTask={jest.fn()}
                cancelEditingTask={jest.fn()}
                deleteTask={deleteTask}
                changeEditingTask={jest.fn()}
                updateTask={jest.fn()}
                isEditing={false}
            />,
        );

        clickButton(component, 'delete-task-button');

        expectedCallbackWasCalledOnceWith(deleteTask, task.id);
    });

    it('After clicking the button to update a task, callback(updateTask) is called out once', () => {
        const task = { id: 2, task: 'Go running! Update' };
        const updateTask = jest.fn();
        const component = ReactTestUtils.renderIntoDocument(
            <Task
                task={task}
                editTask={jest.fn()}
                cancelEditingTask={jest.fn()}
                deleteTask={jest.fn()}
                changeEditingTask={jest.fn()}
                updateTask={updateTask}
                isEditing={true}
            />,
        );

        clickButton(component, 'update-task-button');

        expectedCallbackWasCalledOnceWith(updateTask, task.id, task.task);
    });
});
