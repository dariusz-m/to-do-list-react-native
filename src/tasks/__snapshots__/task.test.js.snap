// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task tests Task component renders correctly with editing mode 1`] = `
<li
  className="list-group-item"
>
  <button
    className="btn btn-danger float-right"
    id="delete-task-button"
    onClick={[Function]}
  >
    Delete
  </button>
  <input
    className="pr-3"
    onChange={[Function]}
    value="Go running!"
  />
  <button
    className="btn btn-primary float-right mr-2"
    onClick={[Function]}
  >
    Cancel
  </button>
  <button
    className="btn btn-warning float-right mr-2"
    id="update-task-button"
    onClick={[Function]}
  >
    Update
  </button>
</li>
`;

exports[`Task tests Task component renders correctly without editing mode 1`] = `
<li
  className="list-group-item"
>
  <button
    className="btn btn-danger float-right"
    id="delete-task-button"
    onClick={[Function]}
  >
    Delete
  </button>
  <p
    className="pr-3"
  >
    Go running!
  </p>
  <button
    className="btn btn-info float-right mr-2"
    onClick={[Function]}
  >
    Edit
  </button>
</li>
`;
